[SherisseTJW] iP

SherisseTJW opened the pull request:
**Professor Oak**

> Professor Oak is your life-long guide and companion. He acts as your 'task-dex' to free your mind of having to remember things you need to do. 

It's,

text-based
easy to learn
FAST SUPER FAST to use
All you need to do is,

download it from [here](https://github.com/SherisseTJW/ip/releases).
add your tasks.
let it manage your tasks for you 😉
And it is FREE!

Features:

- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders
- [ ] GUI

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

SherisseTJW committed:
feat: Level 0. Rename, Greet, Exit

SherisseTJW committed:
feat: Level-1. Echo

SherisseTJW committed:
feat: Level-2. Add, List

SherisseTJW committed:
wip: Level-3

Main functionality complete, clean up remaining

SherisseTJW committed:
wip: Minor Cleanup, some bugs left

SherisseTJW committed:
feat: Level-3

SherisseTJW committed:
feat: Level-4

Added 3 different types of Tasks: Todo, Event and Deadline;

Refactored the code to make it cleaner

SherisseTJW committed:
wip: Text-UI Testing

There is some wierd bug with the OAK ASCII Art, but everything else
seems fine

SherisseTJW committed:
feat: Add exception handling

SherisseTJW committed:
feat: Add delete task

SherisseTJW committed:
test: Add flag to ignore whitespaces in .sh file

SherisseTJW committed:
test: Add test files for exception handling and delete

SherisseTJW committed:
wip: Update TaskService to store tasks

To fulfill Level-7 requirements, we have to allow for the TaskService to
be able to read from and load, as well as save to the tasklist.txt file

This commit is a work-in-progress commit, and only adds 2 empty files:
1. src/main/java/resources/tasklist.txt
2. src/main/java/service/dbService.java

as well as adding an incomplete private method to TaskService.json
to load tasks from tasklist.txt

SherisseTJW committed:
feat: Allow for storage of tasks in separate file

Tasks are not persistent across multiple sessions

To improve the user experience and maintain tasks across multiple
sessions, we have to allow for the tasks to be stored somewhere such
that it can be retrieved in future sessions and is constantly updated as
the user adds / removes tasks

A tasklist.txt has been introduced to maintain these tasks and multiple
helper functions have been added to handle the parsing of the
tasklist.txt as well as the updating of the tasklist.txt whenever the
user makes any changes

SherisseTJW committed:
Merge branch 'branch-Level-7'

SherisseTJW committed:
wip: Add support for Dates and Times

Works for User Input parsing and saving but need to add support for retrieving the task datetime information from tasklist.txt

SherisseTJW committed:
feat: Add support for dates and times

The user can provide any string of characters as the date and time of
events and deadlines.

While this simplifies the parsing of user inputs and reduces the code
complexity, there is no checking done regarding the date time and
neither is there a standard format for the user

We add support for the parsing of the user input for the date and times
of events and deadlines, formatting them in the toString() method of
each task type model.

However, it currently relies on the user following a fixed format so as
to avoid DateTimeParseExceptions when parsing the user input which
should be fixed if we would like the user to have greater flexibility
and to reduce the number of false-positive exceptions occurring

SherisseTJW committed:
Merge branch 'branch-Level-8'

SherisseTJW committed:
Merge remote-tracking branch 'origin/add-gradle-support'

SherisseTJW committed:
wip: Adding JUnit Tests

SherisseTJW committed:
package: Add Mockito Package to gradle build file

SherisseTJW committed:
wip: Refactoring folder structure

SherisseTJW committed:
refactor: Refactor folder structure

The way the folders were originally structured, there was a lot of
coupling between the different packages, etc.

I refactored the folders to contain the purpose rather than the type of
the files within. I.e., rather than having both FeedbackService and
TaskService in /Service, I created two folders, Feedback and Task, to
contain all files related to them in the two folders.

This should reduce coupling between the different files

SherisseTJW committed:
docs: Add java docs

SherisseTJW committed:
doc: Add java docs for fields

SherisseTJW committed:
fix: Mark and unmark task now save in tasklsit.txt

A bug was introduced in the master branch when the feature to support
storing tasks in tasklist.txt to ensure persistence was added. Marking
and unmarking of tasks did not update in tasklist.txt, resulting in
errors in listed tasks in future sessions

An additional utility method in FileUtility is added to support the
updating of existing file contents, which is then called by the mark and
unmark task methods in TaskService

SherisseTJW committed:
minor: Remove completed todo comments

SherisseTJW committed:
Merge branch 'A-CodingStandard'

SherisseTJW committed:
doc: Update README to use OAK ASCII

SherisseTJW committed:
Merge branch 'branch-A-JavaDoc'

SherisseTJW committed:
Merge branch 'master' into feat/A-JUnit

SherisseTJW committed:
test: Add JUnit Tests for Utilities

SherisseTJW committed:
Merge branch 'feat/A-JUnit'

SherisseTJW committed:
feat: Add support for finding tasks

SherisseTJW committed:
doc: Add JavaDoc for the new methods to find tasks

SherisseTJW committed:
Merge branch 'branch-Level-9'

appleraincoat commented on src/main/java/Oak/exceptions/InvalidInputException.java:
```diff
@@ -0,0 +1,39 @@
+        public InvalidCommandException(String invalidCommand) {
+            super("Invalid Command: " + invalidCommand + " received.\n" +
+                    "Only the following commands are accepted:\n" +
+                    Arrays.toString(CommandEnum.values()));
```
I like your usage of line wrapping here!

appleraincoat commented on src/main/java/Oak/feedback/FeedbackService.java:
```diff
@@ -0,0 +1,206 @@
+                String[] eventValues = this.parseEventInput(cur);
+                feedback = this.taskService.addEvent(eventValues[0], eventValues[1], eventValues[2]);
+                break;
+            case Delete:
```
Perhaps you can consider making the case clauses be on the same indentation level as the switch statement?

SherisseTJW replied:
fixed! ^ ^

appleraincoat commented on src/main/java/Oak/feedback/enums/CommandEnum.java:
```diff
@@ -0,0 +1,35 @@
+    Delete("delete"),
+    Todo("todo"),
+    Deadline("deadline"),
+    Event("event");
```
Consider making all enums fully capitalised to follow convention 😃

SherisseTJW replied:
fixed!

appleraincoat commented on src/main/java/Oak/feedback/enums/CommandEnum.java:
```diff
@@ -0,0 +1,35 @@
+
+    // @@author SherisseTJW-reused
+    // Reused from https://www.baeldung.com/java-search-enum-values, Section 3. Searching an Enum by Value
+    // with minor modifications
```
I like how you remembered to indicate areas of code re-use throughout your project. Great work!

appleraincoat reviewed (commented):
Besides some minor nits with the switches,  everything else looks great.  Method and variable names look good to me, and you adhered to appropriate casing. Nice job!

faaheem13 commented on src/main/java/Oak/controller/TerminalController.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.Scanner;
+
+/**
+ * The type Terminal controller.
```
Could the class description be more descriptive, maybe mention what the terminal controller does?

SherisseTJW replied:
fixed!

faaheem13 commented on src/main/java/Oak/task/TaskService.java:
```diff
@@ -0,0 +1,238 @@
+    /** An Array of all the tasks in the system */
+    private ArrayList<Task> tasks = new ArrayList<>();
+    /** The relative file path to the tasklist.txt where all the tasks are stored  */
+    private final String tasklistFilePath = "/src/main/java/Oak/resources/tasklist.txt";
```
i like how you've described each variable and its purpose!

faaheem13 reviewed (commented):
LGTM! Code is extremely easy to read! Clean and kept simple!

SherisseTJW committed:
refactor: Clean up code according to PR

SherisseTJW committed:
style: Refactor according to checkstyle

SherisseTJW committed:
Merge branch 'A-Varargs'

SherisseTJW committed:
wip: Following javafx guide

Morninggloryyy commented on src/main/java/oak/controller/TerminalController.java:
```diff
@@ -0,0 +1,53 @@
+     */
+    private void listen() {
+        Scanner scanner = new Scanner(System.in);
+        boolean stop = false;
```
I guess it might be better to use another name like 'isStop'? (referred to "Boolean variables/methods should be named to sound like booleans" in Java coding standard) :))

Morninggloryyy commented on src/main/java/oak/utility/DateTimeUtility.java:
```diff
@@ -0,0 +1,43 @@
+        // NOTE: We expect the user-input format to be: "<date> @ <time>"
+        String[] rawDateTimeArr = rawDateTime.split("@");
+        LocalDate date = LocalDate.parse(rawDateTimeArr[0].strip());;
+        LocalTime time = rawDateTimeArr.length > 1 ? LocalTime.parse(rawDateTimeArr[1].strip()) : LocalTime.MIDNIGHT;
```
I guess it might be better to try to keep line length shorter than 110 characters. ^^

Morninggloryyy commented on src/test/java/oak/utility/DateTimeUtilityTest.java:
```diff
@@ -0,0 +1,39 @@
+
+public class DateTimeUtilityTest {
+    @Test
+    public void parseValidUserInputStringToLocalDateTimeTest() {
```
I think this name is clear, but maybe it can be a bit shorter?

Morninggloryyy reviewed (commented):
LGTM! I think it might be better to make some methods' names a bit shorter. :))

SherisseTJW committed:
wip: Adding gui support

SherisseTJW committed:
refactor: Folder Structure to account for javafx

SherisseTJW committed:
feat: Add close stage support

SherisseTJW committed:
Merge branch 'Level-10'

SherisseTJW committed:
feat: Add Assertions in TaskService

No assertions in iP Project at all

Assertions help to define assumptions about the program state so we can
verify the program and reduce bugs

Add assertions in the Task Service to check that tasks are stored and
updated correctly

SherisseTJW committed:
feat: Add support for reminders on startup

No reminders are provided to user at all, no command support as well

Given that the user may have a large number of tasks and no sorting of
tasks by deadline is done, the user may not be able to accurately
remember and search for tasks which are due soon

We add support for reminders to be sent to user on startup by:
- Checking all currently available tasks for their by / to datetime
- Have Oak also print reminders on startup

Possible Additional Support:
- Create a 'Reminder' command to get all reminders even after inital
  startup reminder

SherisseTJW committed:
feat: Add Reminder Command Support

Reminders are only given upon startup

The user may want reminders while using the application, i.e., if they
keep the application open throughout instead of only starting it when
requried

A Reminder Command is introduced using the keyword `reminder` which will
print all tasks reminders. This is a similar method of interaction as
that of the other commands, making it more familiar and user
friendly

SherisseTJW committed:
Merge pull request #1 from SherisseTJW/branch-B-Reminders

BCD-Extensions: B-Reminders

SherisseTJW committed:
Merge branch 'master' into branch-A-Assertions

SherisseTJW committed:
Merge pull request #2 from SherisseTJW/branch-A-Assertions

feat: Add Assertions in TaskService

SherisseTJW committed:
wip-doc: Update User Guide

SherisseTJW committed:
doc: Update User Guide

SherisseTJW committed:
feat: Update GUI to resize horizontally

SherisseTJW committed:
doc: Finalise User Guide

SherisseTJW committed:
refactor: Clean code and add javaDocs comments

SherisseTJW committed:
minor: Add comments and clean

SherisseTJW committed:
doc: User Guide for Submission

SherisseTJW committed:
style: Clean all Code Quality Issues

SherisseTJW committed:
Merge pull request #4 from SherisseTJW/branch-A-CodeQuality

A-CodeQuality

SherisseTJW committed:
fix: Truncated Text

Fix obtained from: https://github.com/Fidget-Spinner/ip/commit/6a3f91b3e76d92bb386f83f11d00ee4d0d9cc7c4

SherisseTJW committed:
minor: Remove maxWidth

SherisseTJW committed:
fix: Issues with tasklist.txt and UI

SherisseTJW committed:
feat: Enhancements

SherisseTJW committed:
feat: Add a catch-all OakException Class for any issues

SherisseTJW committed:
style: Fix Code Style errors as per Automated Analysis

https://github.com/SherisseTJW/ip/issues/5
